{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "S9yoZpr2ZOtj",
        "outputId": "a2a6e3a3-2ee1-4f06-caeb-b593b22e3eda"
      },
      "outputs": [],
      "source": [
        "from IPython.core.interactiveshell import InteractiveShell\n",
        "InteractiveShell.ast_node_interactivity = \"all\"\n",
        "%load_ext autoreload\n",
        "%autoreload 2\n",
        "import numpy as np\n",
        "import pandas as pd"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "id": "1CJ8m6icNcBL"
      },
      "outputs": [],
      "source": [
        "from itertools import combinations\n",
        "from scipy.stats import hypergeom\n",
        "import math\n",
        "\n",
        "class CoCo:\n",
        "    \"\"\"Class to calculate the CoCo Score for a given gene-set in a network.\"\"\"\n",
        "    def __init__(self, total_nodes, total_interactions, total_triangles, local_nodes, local_interactions, local_triangles):\n",
        "        self.Nint = total_nodes * (total_nodes - 1) / 2\n",
        "        self.Kint = total_interactions\n",
        "        self.nint = local_nodes * (local_nodes - 1) / 2\n",
        "        self.kint = local_interactions\n",
        "        self.Ntri = total_nodes * (total_nodes - 1) * (total_interactions - 2) / 6\n",
        "        self.Ktri = total_triangles\n",
        "        self.ntri = local_nodes * (local_nodes - 1) * (local_nodes - 2) / 6\n",
        "        self.ktri = local_triangles\n",
        "\n",
        "    def _compute_score(self, k, n, K, N):\n",
        "        if k == 0 or n == 0:\n",
        "            return 0\n",
        "        sign = (k / n) - (K / N)\n",
        "        p = hypergeom.pmf(k, N, K, n)\n",
        "        return 0 if p == 0 else abs(math.log10(p)) * (-1 if sign < 0 else 1)\n",
        "\n",
        "    @property\n",
        "    def CoI(self):\n",
        "        return self._compute_score(self.kint, self.nint, self.Kint, self.Nint)\n",
        "\n",
        "    @property\n",
        "    def CoT(self):\n",
        "        return self._compute_score(self.ktri, self.ntri, self.Ktri, self.Ntri)\n",
        "\n",
        "    @property\n",
        "    def coco(self):\n",
        "        return self.CoI + self.CoT\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "class WeightedGraph:\n",
        "    def __init__(self):\n",
        "        self.graph = {}\n",
        "\n",
        "    def node_exists(self, node):\n",
        "        \"\"\"Check if a node exists in the graph.\"\"\"\n",
        "        return node in self.graph\n",
        "\n",
        "    def edge_exists(self, node1, node2):\n",
        "        \"\"\"Check if an edge exists between node1 and node2.\"\"\"\n",
        "        return self.node_exists(node1) and node2 in self.graph[node1]\n",
        "\n",
        "    def add_node(self, node):\n",
        "        \"\"\"Add a node to the graph.\"\"\"\n",
        "        if not self.node_exists(node):\n",
        "            self.graph[node] = {}\n",
        "\n",
        "    def add_edge(self, node1, node2, weight):\n",
        "        \"\"\"Add an edge between node1 and node2 with the given weight.\"\"\"\n",
        "        if not self.edge_exists(node1, node2):\n",
        "            # If the nodes don't exist, add them to the graph\n",
        "            self.add_node(node1)\n",
        "            self.add_node(node2)\n",
        "\n",
        "            # Add node2 as a neighbor of node1 and vice versa\n",
        "            self.graph[node1][node2] = float(weight)\n",
        "            # self.graph[node2][node1] = weight  # Assuming it's an undirected graph\n",
        "\n",
        "    @property\n",
        "    def get_triangles_count(self):\n",
        "        triangles = 0\n",
        "\n",
        "        for node in self.graph:\n",
        "            neighbors = list(self.graph[node].keys())\n",
        "            for i in range(len(neighbors)):\n",
        "                for j in range(i+1, len(neighbors)):\n",
        "                    if self.edge_exists(neighbors[i], neighbors[j]):\n",
        "                        triangles += 1\n",
        "\n",
        "        # Each triangle is counted 3 times (once for each node), so divide by 3\n",
        "        return triangles // 3\n",
        "\n",
        "    @property\n",
        "    def get_edges_count(self, threshold=0.45):\n",
        "        total_edges = sum(1 for neighbors in self.graph.values() for weight in neighbors.values() if weight > threshold)\n",
        "        return total_edges // 2\n",
        "\n",
        "\n",
        "\n",
        "    def __str__(self):\n",
        "        return str(self.graph)\n",
        "\n",
        "def count_edges_in_subgraph(graph, nodes):\n",
        "    edge_count = 0\n",
        "\n",
        "    # Check each pair of nodes\n",
        "    for i in range(len(nodes)):\n",
        "        for j in range(i+1, len(nodes)):\n",
        "            if nodes[j] in graph[nodes[i]]:\n",
        "                edge_count += 1\n",
        "\n",
        "    return edge_count\n",
        "\n",
        "def count_triangles_in_subgraph(graph, nodes):\n",
        "    triangle_count = 0\n",
        "\n",
        "    # Check every combination of three nodes\n",
        "    for i in range(len(nodes)):\n",
        "        for j in range(i+1, len(nodes)):\n",
        "            for k in range(j+1, len(nodes)):\n",
        "                node1, node2, node3 = nodes[i], nodes[j], nodes[k]\n",
        "                if (node2 in graph[node1] and node3 in graph[node1] and\n",
        "                    node3 in graph[node2]):\n",
        "                    triangle_count += 1\n",
        "\n",
        "    return triangle_count\n",
        "\n",
        "def find_max_coco_subgraph(graph):  # Added threshold parameter with default value 0\n",
        "    nodes = list(graph.graph.keys())\n",
        "    total_nodes = len(nodes)\n",
        "    total_interactions = graph.get_edges_count if hasattr(graph, 'get_edges_count') else 0\n",
        "    # For simplicity, let's assume the graph has a method to compute total triangles\n",
        "    total_triangles = graph.get_triangles_count if hasattr(graph, 'get_triangles_count') else 0\n",
        "\n",
        "    max_score = -float('inf')\n",
        "    best_subgraph = None\n",
        "\n",
        "    # For simplicity, let's check all subgraphs of size 3\n",
        "    for subgraph_nodes in combinations(nodes, 5):\n",
        "        local_nodes = len(subgraph_nodes)\n",
        "\n",
        "        # Count edges in the subgraph using the provided function\n",
        "        local_interactions = count_edges_in_subgraph(graph.graph, subgraph_nodes)\n",
        "\n",
        "        # Count triangles in the subgraph using the provided function\n",
        "        local_triangles = count_triangles_in_subgraph(graph.graph, subgraph_nodes)\n",
        "\n",
        "        coco = CoCo(total_nodes, total_interactions, total_triangles, local_nodes, local_interactions, local_triangles)\n",
        "        if coco.coco > max_score:\n",
        "            max_score = coco.coco\n",
        "            best_subgraph = subgraph_nodes\n",
        "            print(f'best sub: {best_subgraph}, max coco: {max_score}')\n",
        "\n",
        "    return best_subgraph, max_score\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "ejPzXrz8ZOtk"
      },
      "outputs": [],
      "source": [
        "############################################\n",
        "### PAGER functions for PAGER server API ###\n",
        "############################################\n",
        "import requests\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "class PAGER():\n",
        "    def __init__(self):\n",
        "        self.params = {}\n",
        "\n",
        "    ######\n",
        "    # run_pager is a function connected to PAGER api to perform hypergeometric test and retrieve enriched PAGs associated to a list of genes\n",
        "    # The input parameters are:\n",
        "    # 1.genes: a list of gene symbols.\n",
        "    # 2.source: a list of sources refered to 'http://discovery.informatics.uab.edu/PAGER/index.php/pages/help'\n",
        "    # 3.Type: a list of PAG types consisting of 'P', 'A' or 'G'\n",
        "    # 4.minSize: the allowed minimum size of PAG genes\n",
        "    # 5.maxSize: the allowed maximum size of PAG genes\n",
        "    # 6.similarity: the similarity score cutoff [0,1]\n",
        "    # 7.overlap: the allowed minimum overlap genes\n",
        "    # 8.nCoCo: the minimum nCoCo score\n",
        "    # 9.pValue: p-value cutoff\n",
        "    # 10.FDR: false discovery rate\n",
        "    ######\n",
        "    def run_pager(self, genes,**kwargs):\n",
        "        source = kwargs['source'] if 'source' in kwargs.keys() else ['KEGG_2021_HUMAN', 'WikiPathway_2021', 'BioCarta', 'Reactome_2021', 'Spike']\n",
        "        Type = kwargs['Type'] if 'Type' in kwargs.keys() else 'All'\n",
        "        minSize = kwargs['minSize'] if 'minSize' in kwargs.keys() else 1\n",
        "        maxSize = kwargs['maxSize'] if 'maxSize' in kwargs.keys() else 2000\n",
        "        sim = kwargs['similarity'] if 'similarity' in kwargs.keys() else 0\n",
        "        overlap = kwargs['overlap'] if 'overlap' in kwargs.keys() else '1'\n",
        "        organism = kwargs['organism'] if 'organism' in kwargs.keys() else 'All'\n",
        "        nCoCo = kwargs['nCoCo'] if 'nCoCo' in kwargs.keys() else '0'\n",
        "        pvalue = kwargs['pvalue'] if 'pvalue' in kwargs.keys() else 0.05\n",
        "        FDR = kwargs['FDR'] if 'FDR' in kwargs.keys() else 0.05\n",
        "        # Set up the call parameters as a dict.\n",
        "        params = {}\n",
        "        # Work around PAGER API form encode issue.\n",
        "        params['genes'] = '%20'.join(genes)\n",
        "        params['source'] = '%20'.join(source)\n",
        "        params['type'] = Type\n",
        "        params['ge'] = minSize\n",
        "        params['le'] = maxSize\n",
        "        params['sim'] = str(sim)\n",
        "        params['olap'] = str(overlap)\n",
        "        params['organism'] = organism\n",
        "        params['cohesion'] = str(nCoCo)\n",
        "        params['pvalue'] = pvalue\n",
        "        params['FDR'] = FDR\n",
        "        response = requests.post('http://discovery.informatics.uab.edu/PAGER/index.php/geneset/pagerapi', data=params)\n",
        "        #print(response.request.body)\n",
        "        return pd.DataFrame(response.json())\n",
        "\n",
        "    # pathMember is a function connected to PAGER api to retrieve the membership of PAGs using a list of PAG IDs\n",
        "    def pathMember(self, PAG_IDs):\n",
        "        # Set up the call parameters as a dict.\n",
        "        params = {}\n",
        "        params['pag'] = ','.join(PAG_IDs)\n",
        "        # Work around PAGER API form encode issue.\n",
        "        response = requests.post('http://discovery.informatics.uab.edu/PAGER/index.php/geneset/get_members_by_ids/', data=params)\n",
        "        #print(response.request.body)\n",
        "        return pd.DataFrame(response.json()['data'])\n",
        "\n",
        "    # pathInt is a function connected to PAGER api to retrieve the m-type relationships of PAGs using a list of PAG IDs\n",
        "    def pathInt(self, PAG_IDs):\n",
        "        # Set up the call parameters as a dict.\n",
        "        params = {}\n",
        "        params['pag'] = ','.join(PAG_IDs)\n",
        "        # Work around PAGER API form encode issue.\n",
        "        response = requests.post('http://discovery.informatics.uab.edu/PAGER/index.php/pag_pag/inter_network_int_api/', data=params)\n",
        "        #print(response.request.body)\n",
        "        return pd.DataFrame(response.json()['data'])\n",
        "\n",
        "    # pathReg is a function connected to PAGER api to retrieve the r-type relationships of PAGs using a list of PAG IDs\n",
        "    def pathReg(self, PAG_IDs):\n",
        "        # Set up the call parameters as a dict.\n",
        "        params = {}\n",
        "        params['pag'] = ','.join(PAG_IDs)\n",
        "        # Work around PAGER API form encode issue.\n",
        "        response = requests.post('http://discovery.informatics.uab.edu/PAGER/index.php/pag_pag/inter_network_reg_api/', data=params)\n",
        "        #print(response.request.body)\n",
        "        return pd.DataFrame(response.json()['data'])\n",
        "\n",
        "    # pagRankedGene is a function connected to PAGER api to retrieve RP-ranked genes with RP-score of the given PAG_IDs\n",
        "    def pagRankedGene(self, PAGid):\n",
        "        response = requests.get('http://discovery.informatics.uab.edu/PAGER/index.php/genesinPAG/viewgenes/'+PAGid)\n",
        "        return pd.DataFrame(response.json()['gene'])\n",
        "\n",
        "    # pagGeneInt is a function connected to PAGER api to retrieve gene interaction network\n",
        "    def pagGeneInt(self, PAGid):\n",
        "        response = requests.get('http://discovery.informatics.uab.edu/PAGER/index.php/pag_mol_mol_map/interactions/'+PAGid)\n",
        "        return pd.DataFrame(response.json()['data'])\n",
        "\n",
        "    # pagGeneReg is a function connected to PAGER api to retrieve gene regulatory network\n",
        "    def pagGeneReg(self, PAGid):\n",
        "        response = requests.get('http://discovery.informatics.uab.edu/PAGER/index.php/pag_mol_mol_map/regulations/'+PAGid)\n",
        "        return pd.DataFrame(response.json()['data'])\n",
        "\n",
        "    # path_NGSEA is a function connected to PAGER api to generate the network-based GSEA result\n",
        "    import requests\n",
        "    def path_NGSEA(SELF, genes, PAGmember):\n",
        "        geneExpStr = ''\n",
        "        for rowIdx in range(0,genes.shape[0]):\n",
        "            geneExpStr = geneExpStr + genes.iloc[rowIdx,0]+'\\\\t\\\\t'+genes.iloc[rowIdx,1] + \"\\\\t\\\\t\\\\t\"\n",
        "        PAGsetsStr = ''\n",
        "        for rowIdx in range(0,PAGmember.shape[0]):\n",
        "            PAGsetsStr = PAGsetsStr + PAGmember.iloc[rowIdx,0]+'\\\\t\\\\t'+PAGmember.iloc[rowIdx,1] + \"\\\\t\\\\t\\\\t\"\n",
        "        params = {}\n",
        "        params['geneExpStr'] = geneExpStr\n",
        "        params['PAGsetsStr'] = PAGsetsStr\n",
        "        response = requests.post('http://discovery.informatics.uab.edu/PAGER/index.php/geneset/ngseaapi/', data=params)\n",
        "        #print(response.request.body)\n",
        "        return pd.DataFrame(response.json()['data'])\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "id": "nNfoLUXsZOtl"
      },
      "outputs": [],
      "source": [
        "PAGER = PAGER()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "dZZkrhemZOtl",
        "outputId": "c6f7ff66-d801-439b-fcdc-d74eaf5c75a2"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>GS_ID</th>\n",
              "      <th>NAME</th>\n",
              "      <th>SOURCE</th>\n",
              "      <th>GS_SIZE</th>\n",
              "      <th>ORGANISM</th>\n",
              "      <th>DESCRIPTION</th>\n",
              "      <th>LINK</th>\n",
              "      <th>TYPE</th>\n",
              "      <th>MULTI_N</th>\n",
              "      <th>OLAP</th>\n",
              "      <th>COCO_V2</th>\n",
              "      <th>SIMILARITY_SCORE</th>\n",
              "      <th>pvalue</th>\n",
              "      <th>Rank</th>\n",
              "      <th>pFDR</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>WAG002693</td>\n",
              "      <td>Sandbox pathway test</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>5</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Sandbox pathway test</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>188.558608</td>\n",
              "      <td>.52</td>\n",
              "      <td>1.020749e-08</td>\n",
              "      <td>1</td>\n",
              "      <td>0.000001</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>WAG000644</td>\n",
              "      <td>role of brca1 brca2 and atr in cancer suscepti...</td>\n",
              "      <td>BioCarta</td>\n",
              "      <td>21</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BRCA1 and BRCA2 were identified genetically as...</td>\n",
              "      <td>https://cgap.nci.nih.gov/Pathways/BioCarta/h_a...</td>\n",
              "      <td>P</td>\n",
              "      <td>252</td>\n",
              "      <td>2</td>\n",
              "      <td>5956.815769</td>\n",
              "      <td>.2761904761904761904761904761904761904762</td>\n",
              "      <td>2.143572e-07</td>\n",
              "      <td>2</td>\n",
              "      <td>0.000024</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>WAG004117</td>\n",
              "      <td>BCDX2 complex stabilizes RAD51 filament</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>22</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BCDX2 complex stabilizes RAD51 filament</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3370.353336</td>\n",
              "      <td>.2727272727272727272727272727272727272727</td>\n",
              "      <td>2.357930e-07</td>\n",
              "      <td>3</td>\n",
              "      <td>0.000026</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>WAG004154</td>\n",
              "      <td>D-loop formation mediated by PALB2, BRCA2 and ...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>24</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>D-loop formation mediated by PALB2, BRCA2 and ...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3436.072517</td>\n",
              "      <td>.2666666666666666666666666666666666666667</td>\n",
              "      <td>2.817267e-07</td>\n",
              "      <td>4</td>\n",
              "      <td>0.000031</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>WAG004162</td>\n",
              "      <td>Defective D-loop formation mediated by PALB2, ...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>24</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Defective D-loop formation mediated by PALB2, ...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3436.072517</td>\n",
              "      <td>.2666666666666666666666666666666666666667</td>\n",
              "      <td>2.817267e-07</td>\n",
              "      <td>4</td>\n",
              "      <td>0.000030</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>WAG004163</td>\n",
              "      <td>Defective D-loop formation mediated by PALB2, ...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>24</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Defective D-loop formation mediated by PALB2, ...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3436.072517</td>\n",
              "      <td>.2666666666666666666666666666666666666667</td>\n",
              "      <td>2.817267e-07</td>\n",
              "      <td>4</td>\n",
              "      <td>0.000030</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>WAG004164</td>\n",
              "      <td>Defective D-loop formation mediated by PALB2, ...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>24</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Defective D-loop formation mediated by PALB2, ...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3436.072517</td>\n",
              "      <td>.2666666666666666666666666666666666666667</td>\n",
              "      <td>2.817267e-07</td>\n",
              "      <td>4</td>\n",
              "      <td>0.000030</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>WAG004118</td>\n",
              "      <td>CX3 complex binds D-loop structures</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>25</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>CX3 complex binds D-loop structures</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3687.288396</td>\n",
              "      <td>.264</td>\n",
              "      <td>3.062246e-07</td>\n",
              "      <td>8</td>\n",
              "      <td>0.000032</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>WAG004123</td>\n",
              "      <td>Resolution of D-loops cleaved by MUS81:EME1 or...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>25</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Resolution of D-loops cleaved by MUS81:EME1 or...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3687.288396</td>\n",
              "      <td>.264</td>\n",
              "      <td>3.062246e-07</td>\n",
              "      <td>8</td>\n",
              "      <td>0.000032</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>WAG004124</td>\n",
              "      <td>Resolution of Holliday junctions cleaved by GE...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>25</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Resolution of Holliday junctions cleaved by GE...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3687.288396</td>\n",
              "      <td>.264</td>\n",
              "      <td>3.062246e-07</td>\n",
              "      <td>8</td>\n",
              "      <td>0.000032</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>WAG004136</td>\n",
              "      <td>Ligation of DNA and formation of Holliday stru...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>25</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Ligation of DNA and formation of Holliday stru...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3687.288396</td>\n",
              "      <td>.264</td>\n",
              "      <td>3.062246e-07</td>\n",
              "      <td>8</td>\n",
              "      <td>0.000031</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>WAG004121</td>\n",
              "      <td>BLM mediates dissolution of double Holliday ju...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>26</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BLM mediates dissolution of double Holliday ju...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3402.224081</td>\n",
              "      <td>.2615384615384615384615384615384615384615</td>\n",
              "      <td>3.317434e-07</td>\n",
              "      <td>12</td>\n",
              "      <td>0.000034</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>WAG004147</td>\n",
              "      <td>D-loop dissociation and strand annealing</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>26</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>D-loop dissociation and strand annealing</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3402.224081</td>\n",
              "      <td>.2615384615384615384615384615384615384615</td>\n",
              "      <td>3.317434e-07</td>\n",
              "      <td>12</td>\n",
              "      <td>0.000033</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>WAG003221</td>\n",
              "      <td>Homologous recombination</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>28</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Homologous recombination</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>1940.666302</td>\n",
              "      <td>.2571428571428571428571428571428571428571</td>\n",
              "      <td>3.858430e-07</td>\n",
              "      <td>14</td>\n",
              "      <td>0.000038</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>WAG004122</td>\n",
              "      <td>MUS81:EME1,EME2 cleaves D-loop</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>28</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>MUS81:EME1,EME2 cleaves D-loop</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>3336.772994</td>\n",
              "      <td>.2571428571428571428571428571428571428571</td>\n",
              "      <td>3.858430e-07</td>\n",
              "      <td>14</td>\n",
              "      <td>0.000038</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>WAG004146</td>\n",
              "      <td>Cleavage of Holliday junctions by GEN1 or SLX1...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>32</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Cleavage of Holliday junctions by GEN1 or SLX1...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>2709.154452</td>\n",
              "      <td>.25</td>\n",
              "      <td>5.062914e-07</td>\n",
              "      <td>16</td>\n",
              "      <td>0.000049</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>WAG004141</td>\n",
              "      <td>RAD51 binds BRCA2 at resected DNA DSBs</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>34</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>RAD51 binds BRCA2 at resected DNA DSBs</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>6125.62014</td>\n",
              "      <td>.2470588235294117647058823529411764705882</td>\n",
              "      <td>5.726401e-07</td>\n",
              "      <td>17</td>\n",
              "      <td>0.000055</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>WAG001743</td>\n",
              "      <td>hsa03440 Homologous recombination - Homo sapie...</td>\n",
              "      <td>KEGG_2021_HUMAN</td>\n",
              "      <td>41</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Homologous recombination (HR) is essential for...</td>\n",
              "      <td>https://www.genome.jp/pathway/hsa03440</td>\n",
              "      <td>P</td>\n",
              "      <td>345</td>\n",
              "      <td>2</td>\n",
              "      <td>2627.667251</td>\n",
              "      <td>.2390243902439024390243902439024390243902</td>\n",
              "      <td>8.370140e-07</td>\n",
              "      <td>18</td>\n",
              "      <td>0.000080</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>WAG004404</td>\n",
              "      <td>Formation of meiotic single-stranded DNA invas...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>44</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Formation of meiotic single-stranded DNA invas...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-91...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>115.3844923</td>\n",
              "      <td>.2363636363636363636363636363636363636364</td>\n",
              "      <td>9.656283e-07</td>\n",
              "      <td>19</td>\n",
              "      <td>0.000091</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>WAG004402</td>\n",
              "      <td>Formation of meiotic heteroduplex</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>45</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Formation of meiotic heteroduplex</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-91...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>150.0163697</td>\n",
              "      <td>.2355555555555555555555555555555555555556</td>\n",
              "      <td>1.010541e-06</td>\n",
              "      <td>20</td>\n",
              "      <td>0.000094</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>WAG004148</td>\n",
              "      <td>D-loop extension by DNA polymerases</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>48</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>D-loop extension by DNA polymerases</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>2</td>\n",
              "      <td>5990.79972</td>\n",
              "      <td>.2333333333333333333333333333333333333333</td>\n",
              "      <td>1.151405e-06</td>\n",
              "      <td>21</td>\n",
              "      <td>0.000106</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>WAG001745</td>\n",
              "      <td>hsa03460 Fanconi anemia pathway - Homo sapiens...</td>\n",
              "      <td>KEGG_2021_HUMAN</td>\n",
              "      <td>54</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>The Fanconi anemia pathway is required for the...</td>\n",
              "      <td>https://www.genome.jp/pathway/hsa03460</td>\n",
              "      <td>P</td>\n",
              "      <td>345</td>\n",
              "      <td>2</td>\n",
              "      <td>2639.032014</td>\n",
              "      <td>.2296296296296296296296296296296296296296</td>\n",
              "      <td>1.460692e-06</td>\n",
              "      <td>22</td>\n",
              "      <td>0.000133</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>WAG002675</td>\n",
              "      <td>DNA IR-double strand breaks and cellular respo...</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>55</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>DNA IR-double strand breaks and cellular respo...</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>2707.216113</td>\n",
              "      <td>.2290909090909090909090909090909090909091</td>\n",
              "      <td>1.515812e-06</td>\n",
              "      <td>23</td>\n",
              "      <td>0.000136</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>WAG002700</td>\n",
              "      <td>DNA IR-damage and cellular response via ATR</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>80</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>DNA IR-damage and cellular response via ATR</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>4220.275213</td>\n",
              "      <td>.22</td>\n",
              "      <td>3.225566e-06</td>\n",
              "      <td>24</td>\n",
              "      <td>0.000287</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>WAG002161</td>\n",
              "      <td>Glioblastoma signaling pathways</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>83</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Glioblastoma signaling pathways</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>4976.097877</td>\n",
              "      <td>.219277108433734939759036144578313253012</td>\n",
              "      <td>3.473608e-06</td>\n",
              "      <td>25</td>\n",
              "      <td>0.000306</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>WAG002265</td>\n",
              "      <td>Meiotic recombination</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>88</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Meiotic recombination</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>344.1773602</td>\n",
              "      <td>.2181818181818181818181818181818181818182</td>\n",
              "      <td>3.907426e-06</td>\n",
              "      <td>26</td>\n",
              "      <td>0.000340</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>WIG001822</td>\n",
              "      <td>Repair of Interstrand Crosslinks</td>\n",
              "      <td>Spike</td>\n",
              "      <td>102</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Multiple pathways of DNA repair participate in...</td>\n",
              "      <td>http://www.cs.tau.ac.il/~spike/maps/spike00007...</td>\n",
              "      <td>P</td>\n",
              "      <td>28</td>\n",
              "      <td>2</td>\n",
              "      <td>5740.10268</td>\n",
              "      <td>.2156862745098039215686274509803921568627</td>\n",
              "      <td>5.257877e-06</td>\n",
              "      <td>27</td>\n",
              "      <td>0.000452</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>WAG003110</td>\n",
              "      <td>DNA repair pathways, full network</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>120</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>DNA repair pathways, full network</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>6283.961347</td>\n",
              "      <td>.2133333333333333333333333333333333333333</td>\n",
              "      <td>7.288146e-06</td>\n",
              "      <td>28</td>\n",
              "      <td>0.000619</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>WAG002505</td>\n",
              "      <td>HDR through Homologous Recombination (HRR) or ...</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>132</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>HDR through Homologous Recombination (HRR) or ...</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>2430.640086</td>\n",
              "      <td>.2121212121212121212121212121212121212121</td>\n",
              "      <td>8.825394e-06</td>\n",
              "      <td>29</td>\n",
              "      <td>0.000741</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>WAG001955</td>\n",
              "      <td>hsa05224 Breast cancer - Homo sapiens (human)</td>\n",
              "      <td>KEGG_2021_HUMAN</td>\n",
              "      <td>147</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Breast cancer is the leading cause of cancer d...</td>\n",
              "      <td>https://www.genome.jp/pathway/hsa05224</td>\n",
              "      <td>P</td>\n",
              "      <td>345</td>\n",
              "      <td>2</td>\n",
              "      <td>3701.25203</td>\n",
              "      <td>.210884353741496598639455782312925170068</td>\n",
              "      <td>1.095366e-05</td>\n",
              "      <td>30</td>\n",
              "      <td>0.000909</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>30</th>\n",
              "      <td>WAG002840</td>\n",
              "      <td>Breast cancer pathway</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>155</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Breast cancer pathway</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>3126.072387</td>\n",
              "      <td>.2103225806451612903225806451612903225806</td>\n",
              "      <td>1.218264e-05</td>\n",
              "      <td>31</td>\n",
              "      <td>0.000999</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>31</th>\n",
              "      <td>WAG002125</td>\n",
              "      <td>Integrated breast cancer pathway</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>157</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Integrated breast cancer pathway</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>2</td>\n",
              "      <td>1553.397473</td>\n",
              "      <td>.2101910828025477707006369426751592356688</td>\n",
              "      <td>1.250009e-05</td>\n",
              "      <td>32</td>\n",
              "      <td>0.001013</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>WIG001825</td>\n",
              "      <td>DNA damage response and chromatin remodeling</td>\n",
              "      <td>Spike</td>\n",
              "      <td>157</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>The packaging of eukaryotic chromosomal DNA in...</td>\n",
              "      <td>http://www.cs.tau.ac.il/~spike/maps/spike00010...</td>\n",
              "      <td>P</td>\n",
              "      <td>28</td>\n",
              "      <td>2</td>\n",
              "      <td>1569.958527</td>\n",
              "      <td>.2101910828025477707006369426751592356688</td>\n",
              "      <td>1.250009e-05</td>\n",
              "      <td>32</td>\n",
              "      <td>0.001000</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>WIG001819</td>\n",
              "      <td>Response to DSBs</td>\n",
              "      <td>Spike</td>\n",
              "      <td>164</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>DNA double-strand breaks (DSBs) arise in cells...</td>\n",
              "      <td>http://www.cs.tau.ac.il/~spike/maps/spike00004...</td>\n",
              "      <td>P</td>\n",
              "      <td>28</td>\n",
              "      <td>2</td>\n",
              "      <td>2707.633139</td>\n",
              "      <td>.2097560975609756097560975609756097560976</td>\n",
              "      <td>1.364333e-05</td>\n",
              "      <td>34</td>\n",
              "      <td>0.001078</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>WAG006529</td>\n",
              "      <td>CHEK1 phosphorylates BRCA2</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>2</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>CHEK1 phosphorylates BRCA2</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>126.2105263</td>\n",
              "      <td>.3666666666666666666666666666666666666667</td>\n",
              "      <td>9.036281e-05</td>\n",
              "      <td>35</td>\n",
              "      <td>0.007048</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>35</th>\n",
              "      <td>WAG014925</td>\n",
              "      <td>BRCA1 forms a heterodimer with BARD1</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>2</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BRCA1 forms a heterodimer with BARD1</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>126.2105263</td>\n",
              "      <td>.3666666666666666666666666666666666666667</td>\n",
              "      <td>9.036281e-05</td>\n",
              "      <td>35</td>\n",
              "      <td>0.006958</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>36</th>\n",
              "      <td>WAG014927</td>\n",
              "      <td>Some pathogenic BRCA1 mutants do not bind BARD1</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>2</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Some pathogenic BRCA1 mutants do not bind BARD1</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-96...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>126.2105263</td>\n",
              "      <td>.3666666666666666666666666666666666666667</td>\n",
              "      <td>9.036281e-05</td>\n",
              "      <td>35</td>\n",
              "      <td>0.006868</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>WAG014928</td>\n",
              "      <td>Defective BARD1 does not bind BRCA1</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>2</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Defective BARD1 does not bind BRCA1</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-96...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>126.2105263</td>\n",
              "      <td>.3666666666666666666666666666666666666667</td>\n",
              "      <td>9.036281e-05</td>\n",
              "      <td>35</td>\n",
              "      <td>0.006777</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>38</th>\n",
              "      <td>WAG013979</td>\n",
              "      <td>UBXN1 binds BRCA1:BARD1 heterodimer</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>3</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>UBXN1 binds BRCA1:BARD1 heterodimer</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>384.3965412</td>\n",
              "      <td>.3</td>\n",
              "      <td>1.355411e-04</td>\n",
              "      <td>39</td>\n",
              "      <td>0.010030</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>WAG014929</td>\n",
              "      <td>BAP1 disrupts the BRCA1:BARD1 heterodimer</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>3</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BAP1 disrupts the BRCA1:BARD1 heterodimer</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>134.1614907</td>\n",
              "      <td>.3</td>\n",
              "      <td>1.355411e-04</td>\n",
              "      <td>39</td>\n",
              "      <td>0.009895</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>40</th>\n",
              "      <td>WAG014931</td>\n",
              "      <td>BAP1 binds BRCA1:BARD1 heterodimer</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>3</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BAP1 binds BRCA1:BARD1 heterodimer</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>134.1614907</td>\n",
              "      <td>.3</td>\n",
              "      <td>1.355411e-04</td>\n",
              "      <td>39</td>\n",
              "      <td>0.009759</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>WAG008846</td>\n",
              "      <td>BRCA1 gene expression is inhibited by E2F6</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>4</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BRCA1 gene expression is inhibited by E2F6</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-90...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>756.782076</td>\n",
              "      <td>.26</td>\n",
              "      <td>1.807174e-04</td>\n",
              "      <td>42</td>\n",
              "      <td>0.012831</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>42</th>\n",
              "      <td>WAG014174</td>\n",
              "      <td>PIAS1,4 SUMOylates BRCA1 with SUMO1</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>5</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>PIAS1,4 SUMOylates BRCA1 with SUMO1</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-29...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>1216.805852</td>\n",
              "      <td>.2333333333333333333333333333333333333333</td>\n",
              "      <td>2.258917e-04</td>\n",
              "      <td>43</td>\n",
              "      <td>0.015812</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>43</th>\n",
              "      <td>WAG014173</td>\n",
              "      <td>PIAS1,4 SUMOylates BRCA1 with SUMO2,3</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>6</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>PIAS1,4 SUMOylates BRCA1 with SUMO2,3</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-29...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>1744.991383</td>\n",
              "      <td>.2142857142857142857142857142857142857143</td>\n",
              "      <td>2.710639e-04</td>\n",
              "      <td>44</td>\n",
              "      <td>0.018703</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>44</th>\n",
              "      <td>WAG014930</td>\n",
              "      <td>BRCA1:BARD1 heterodimer autoubiquitinates</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>7</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BRCA1:BARD1 heterodimer autoubiquitinates</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-97...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>1469.189466</td>\n",
              "      <td>.2</td>\n",
              "      <td>3.162341e-04</td>\n",
              "      <td>45</td>\n",
              "      <td>0.021504</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>45</th>\n",
              "      <td>WAG000212</td>\n",
              "      <td>brca1 dependent ub ligase activity</td>\n",
              "      <td>BioCarta</td>\n",
              "      <td>8</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BRCA1 is a breast and ovarian cancer tumor sup...</td>\n",
              "      <td>https://cgap.nci.nih.gov/Pathways/BioCarta/h_b...</td>\n",
              "      <td>P</td>\n",
              "      <td>252</td>\n",
              "      <td>1</td>\n",
              "      <td>2952.437544</td>\n",
              "      <td>.1888888888888888888888888888888888888889</td>\n",
              "      <td>3.614022e-04</td>\n",
              "      <td>46</td>\n",
              "      <td>0.024214</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>WAG004110</td>\n",
              "      <td>BRCA1 binds phosphorylated RBBP8</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>8</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>BRCA1 binds phosphorylated RBBP8</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>1317.698072</td>\n",
              "      <td>.1888888888888888888888888888888888888889</td>\n",
              "      <td>3.614022e-04</td>\n",
              "      <td>46</td>\n",
              "      <td>0.023853</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>47</th>\n",
              "      <td>WAG004152</td>\n",
              "      <td>Initial resection of double-strand break ends</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>8</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Initial resection of double-strand break ends</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>1317.698072</td>\n",
              "      <td>.1888888888888888888888888888888888888889</td>\n",
              "      <td>3.614022e-04</td>\n",
              "      <td>46</td>\n",
              "      <td>0.023491</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>48</th>\n",
              "      <td>WAG003241</td>\n",
              "      <td>SMC1/SMC3 role in DNA damage - Cornelia de Lan...</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>11</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>SMC1/SMC3 role in DNA damage - Cornelia de Lan...</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>1</td>\n",
              "      <td>1793.045634</td>\n",
              "      <td>.1666666666666666666666666666666666666667</td>\n",
              "      <td>4.968944e-04</td>\n",
              "      <td>49</td>\n",
              "      <td>0.031801</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>49</th>\n",
              "      <td>WAG002069</td>\n",
              "      <td>Homologous recombination</td>\n",
              "      <td>WikiPathway_2021</td>\n",
              "      <td>13</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Homologous recombination</td>\n",
              "      <td>https://www.wikipathways.org/index.php/Pathway...</td>\n",
              "      <td>P</td>\n",
              "      <td>1333</td>\n",
              "      <td>1</td>\n",
              "      <td>1401.916365</td>\n",
              "      <td>.1571428571428571428571428571428571428571</td>\n",
              "      <td>5.872123e-04</td>\n",
              "      <td>50</td>\n",
              "      <td>0.036994</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>WAG004340</td>\n",
              "      <td>Resolution of meiotic holliday junction</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>13</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Resolution of meiotic holliday junction</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-91...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>4341.774771</td>\n",
              "      <td>.1571428571428571428571428571428571428571</td>\n",
              "      <td>5.872123e-04</td>\n",
              "      <td>50</td>\n",
              "      <td>0.036407</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>51</th>\n",
              "      <td>WAG004119</td>\n",
              "      <td>EXO1 or DNA2 in complex with BLM or WRN binds ...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>16</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>EXO1 or DNA2 in complex with BLM or WRN binds ...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>2154.832995</td>\n",
              "      <td>.1470588235294117647058823529411764705882</td>\n",
              "      <td>7.226738e-04</td>\n",
              "      <td>52</td>\n",
              "      <td>0.044083</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>52</th>\n",
              "      <td>WAG004120</td>\n",
              "      <td>Long-range resection of DNA DSBs by EXO1 or DNA2</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>16</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Long-range resection of DNA DSBs by EXO1 or DNA2</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>2154.832995</td>\n",
              "      <td>.1470588235294117647058823529411764705882</td>\n",
              "      <td>7.226738e-04</td>\n",
              "      <td>52</td>\n",
              "      <td>0.043360</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53</th>\n",
              "      <td>WAG000628</td>\n",
              "      <td>atm signaling pathway</td>\n",
              "      <td>BioCarta</td>\n",
              "      <td>18</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>The ataxia telangiectasia-mutated gene (ATM) e...</td>\n",
              "      <td>https://cgap.nci.nih.gov/Pathways/BioCarta/h_a...</td>\n",
              "      <td>P</td>\n",
              "      <td>252</td>\n",
              "      <td>1</td>\n",
              "      <td>2452.469682</td>\n",
              "      <td>.1421052631578947368421052631578947368421</td>\n",
              "      <td>8.129713e-04</td>\n",
              "      <td>54</td>\n",
              "      <td>0.047965</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>54</th>\n",
              "      <td>WAG004138</td>\n",
              "      <td>Association of RPA complexes with ssDNA at res...</td>\n",
              "      <td>Reactome_2021</td>\n",
              "      <td>19</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>Association of RPA complexes with ssDNA at res...</td>\n",
              "      <td>https://reactome.org/PathwayBrowser/#/R-HSA-56...</td>\n",
              "      <td>P</td>\n",
              "      <td>13163</td>\n",
              "      <td>1</td>\n",
              "      <td>3183.379618</td>\n",
              "      <td>.14</td>\n",
              "      <td>8.581170e-04</td>\n",
              "      <td>55</td>\n",
              "      <td>0.049771</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>55</th>\n",
              "      <td>WIG001826</td>\n",
              "      <td>WNT signaling</td>\n",
              "      <td>Spike</td>\n",
              "      <td>448</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>The Wnt pathway plays a critical role in the d...</td>\n",
              "      <td>http://www.cs.tau.ac.il/~spike/maps/spike00027...</td>\n",
              "      <td>P</td>\n",
              "      <td>28</td>\n",
              "      <td>1</td>\n",
              "      <td>1198.7612</td>\n",
              "      <td>.1017817371937639198218262806236080178174</td>\n",
              "      <td>2.003731e-02</td>\n",
              "      <td>111</td>\n",
              "      <td>0.040075</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>56</th>\n",
              "      <td>WAG001933</td>\n",
              "      <td>hsa05200 Pathways in cancer - Homo sapiens (hu...</td>\n",
              "      <td>KEGG_2021_HUMAN</td>\n",
              "      <td>531</td>\n",
              "      <td>Homo sapiens</td>\n",
              "      <td>None</td>\n",
              "      <td>https://www.genome.jp/pathway/hsa05200</td>\n",
              "      <td>P</td>\n",
              "      <td>345</td>\n",
              "      <td>1</td>\n",
              "      <td>2740.388822</td>\n",
              "      <td>.1015037593984962406015037593984962406015</td>\n",
              "      <td>2.370460e-02</td>\n",
              "      <td>112</td>\n",
              "      <td>0.023705</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "        GS_ID                                               NAME  \\\n",
              "0   WAG002693                               Sandbox pathway test   \n",
              "1   WAG000644  role of brca1 brca2 and atr in cancer suscepti...   \n",
              "2   WAG004117            BCDX2 complex stabilizes RAD51 filament   \n",
              "3   WAG004154  D-loop formation mediated by PALB2, BRCA2 and ...   \n",
              "4   WAG004162  Defective D-loop formation mediated by PALB2, ...   \n",
              "5   WAG004163  Defective D-loop formation mediated by PALB2, ...   \n",
              "6   WAG004164  Defective D-loop formation mediated by PALB2, ...   \n",
              "7   WAG004118                CX3 complex binds D-loop structures   \n",
              "8   WAG004123  Resolution of D-loops cleaved by MUS81:EME1 or...   \n",
              "9   WAG004124  Resolution of Holliday junctions cleaved by GE...   \n",
              "10  WAG004136  Ligation of DNA and formation of Holliday stru...   \n",
              "11  WAG004121  BLM mediates dissolution of double Holliday ju...   \n",
              "12  WAG004147           D-loop dissociation and strand annealing   \n",
              "13  WAG003221                           Homologous recombination   \n",
              "14  WAG004122                     MUS81:EME1,EME2 cleaves D-loop   \n",
              "15  WAG004146  Cleavage of Holliday junctions by GEN1 or SLX1...   \n",
              "16  WAG004141             RAD51 binds BRCA2 at resected DNA DSBs   \n",
              "17  WAG001743  hsa03440 Homologous recombination - Homo sapie...   \n",
              "18  WAG004404  Formation of meiotic single-stranded DNA invas...   \n",
              "19  WAG004402                  Formation of meiotic heteroduplex   \n",
              "20  WAG004148                D-loop extension by DNA polymerases   \n",
              "21  WAG001745  hsa03460 Fanconi anemia pathway - Homo sapiens...   \n",
              "22  WAG002675  DNA IR-double strand breaks and cellular respo...   \n",
              "23  WAG002700        DNA IR-damage and cellular response via ATR   \n",
              "24  WAG002161                    Glioblastoma signaling pathways   \n",
              "25  WAG002265                              Meiotic recombination   \n",
              "26  WIG001822                   Repair of Interstrand Crosslinks   \n",
              "27  WAG003110                  DNA repair pathways, full network   \n",
              "28  WAG002505  HDR through Homologous Recombination (HRR) or ...   \n",
              "29  WAG001955      hsa05224 Breast cancer - Homo sapiens (human)   \n",
              "30  WAG002840                              Breast cancer pathway   \n",
              "31  WAG002125                   Integrated breast cancer pathway   \n",
              "32  WIG001825       DNA damage response and chromatin remodeling   \n",
              "33  WIG001819                                   Response to DSBs   \n",
              "34  WAG006529                         CHEK1 phosphorylates BRCA2   \n",
              "35  WAG014925               BRCA1 forms a heterodimer with BARD1   \n",
              "36  WAG014927    Some pathogenic BRCA1 mutants do not bind BARD1   \n",
              "37  WAG014928                Defective BARD1 does not bind BRCA1   \n",
              "38  WAG013979                UBXN1 binds BRCA1:BARD1 heterodimer   \n",
              "39  WAG014929          BAP1 disrupts the BRCA1:BARD1 heterodimer   \n",
              "40  WAG014931                 BAP1 binds BRCA1:BARD1 heterodimer   \n",
              "41  WAG008846         BRCA1 gene expression is inhibited by E2F6   \n",
              "42  WAG014174                PIAS1,4 SUMOylates BRCA1 with SUMO1   \n",
              "43  WAG014173              PIAS1,4 SUMOylates BRCA1 with SUMO2,3   \n",
              "44  WAG014930          BRCA1:BARD1 heterodimer autoubiquitinates   \n",
              "45  WAG000212                 brca1 dependent ub ligase activity   \n",
              "46  WAG004110                   BRCA1 binds phosphorylated RBBP8   \n",
              "47  WAG004152      Initial resection of double-strand break ends   \n",
              "48  WAG003241  SMC1/SMC3 role in DNA damage - Cornelia de Lan...   \n",
              "49  WAG002069                           Homologous recombination   \n",
              "50  WAG004340            Resolution of meiotic holliday junction   \n",
              "51  WAG004119  EXO1 or DNA2 in complex with BLM or WRN binds ...   \n",
              "52  WAG004120   Long-range resection of DNA DSBs by EXO1 or DNA2   \n",
              "53  WAG000628                              atm signaling pathway   \n",
              "54  WAG004138  Association of RPA complexes with ssDNA at res...   \n",
              "55  WIG001826                                      WNT signaling   \n",
              "56  WAG001933  hsa05200 Pathways in cancer - Homo sapiens (hu...   \n",
              "\n",
              "              SOURCE GS_SIZE      ORGANISM  \\\n",
              "0   WikiPathway_2021       5  Homo sapiens   \n",
              "1           BioCarta      21  Homo sapiens   \n",
              "2      Reactome_2021      22  Homo sapiens   \n",
              "3      Reactome_2021      24  Homo sapiens   \n",
              "4      Reactome_2021      24  Homo sapiens   \n",
              "5      Reactome_2021      24  Homo sapiens   \n",
              "6      Reactome_2021      24  Homo sapiens   \n",
              "7      Reactome_2021      25  Homo sapiens   \n",
              "8      Reactome_2021      25  Homo sapiens   \n",
              "9      Reactome_2021      25  Homo sapiens   \n",
              "10     Reactome_2021      25  Homo sapiens   \n",
              "11     Reactome_2021      26  Homo sapiens   \n",
              "12     Reactome_2021      26  Homo sapiens   \n",
              "13  WikiPathway_2021      28  Homo sapiens   \n",
              "14     Reactome_2021      28  Homo sapiens   \n",
              "15     Reactome_2021      32  Homo sapiens   \n",
              "16     Reactome_2021      34  Homo sapiens   \n",
              "17   KEGG_2021_HUMAN      41  Homo sapiens   \n",
              "18     Reactome_2021      44  Homo sapiens   \n",
              "19     Reactome_2021      45  Homo sapiens   \n",
              "20     Reactome_2021      48  Homo sapiens   \n",
              "21   KEGG_2021_HUMAN      54  Homo sapiens   \n",
              "22  WikiPathway_2021      55  Homo sapiens   \n",
              "23  WikiPathway_2021      80  Homo sapiens   \n",
              "24  WikiPathway_2021      83  Homo sapiens   \n",
              "25  WikiPathway_2021      88  Homo sapiens   \n",
              "26             Spike     102  Homo sapiens   \n",
              "27  WikiPathway_2021     120  Homo sapiens   \n",
              "28  WikiPathway_2021     132  Homo sapiens   \n",
              "29   KEGG_2021_HUMAN     147  Homo sapiens   \n",
              "30  WikiPathway_2021     155  Homo sapiens   \n",
              "31  WikiPathway_2021     157  Homo sapiens   \n",
              "32             Spike     157  Homo sapiens   \n",
              "33             Spike     164  Homo sapiens   \n",
              "34     Reactome_2021       2  Homo sapiens   \n",
              "35     Reactome_2021       2  Homo sapiens   \n",
              "36     Reactome_2021       2  Homo sapiens   \n",
              "37     Reactome_2021       2  Homo sapiens   \n",
              "38     Reactome_2021       3  Homo sapiens   \n",
              "39     Reactome_2021       3  Homo sapiens   \n",
              "40     Reactome_2021       3  Homo sapiens   \n",
              "41     Reactome_2021       4  Homo sapiens   \n",
              "42     Reactome_2021       5  Homo sapiens   \n",
              "43     Reactome_2021       6  Homo sapiens   \n",
              "44     Reactome_2021       7  Homo sapiens   \n",
              "45          BioCarta       8  Homo sapiens   \n",
              "46     Reactome_2021       8  Homo sapiens   \n",
              "47     Reactome_2021       8  Homo sapiens   \n",
              "48  WikiPathway_2021      11  Homo sapiens   \n",
              "49  WikiPathway_2021      13  Homo sapiens   \n",
              "50     Reactome_2021      13  Homo sapiens   \n",
              "51     Reactome_2021      16  Homo sapiens   \n",
              "52     Reactome_2021      16  Homo sapiens   \n",
              "53          BioCarta      18  Homo sapiens   \n",
              "54     Reactome_2021      19  Homo sapiens   \n",
              "55             Spike     448  Homo sapiens   \n",
              "56   KEGG_2021_HUMAN     531  Homo sapiens   \n",
              "\n",
              "                                          DESCRIPTION  \\\n",
              "0                                Sandbox pathway test   \n",
              "1   BRCA1 and BRCA2 were identified genetically as...   \n",
              "2             BCDX2 complex stabilizes RAD51 filament   \n",
              "3   D-loop formation mediated by PALB2, BRCA2 and ...   \n",
              "4   Defective D-loop formation mediated by PALB2, ...   \n",
              "5   Defective D-loop formation mediated by PALB2, ...   \n",
              "6   Defective D-loop formation mediated by PALB2, ...   \n",
              "7                 CX3 complex binds D-loop structures   \n",
              "8   Resolution of D-loops cleaved by MUS81:EME1 or...   \n",
              "9   Resolution of Holliday junctions cleaved by GE...   \n",
              "10  Ligation of DNA and formation of Holliday stru...   \n",
              "11  BLM mediates dissolution of double Holliday ju...   \n",
              "12           D-loop dissociation and strand annealing   \n",
              "13                           Homologous recombination   \n",
              "14                     MUS81:EME1,EME2 cleaves D-loop   \n",
              "15  Cleavage of Holliday junctions by GEN1 or SLX1...   \n",
              "16             RAD51 binds BRCA2 at resected DNA DSBs   \n",
              "17  Homologous recombination (HR) is essential for...   \n",
              "18  Formation of meiotic single-stranded DNA invas...   \n",
              "19                  Formation of meiotic heteroduplex   \n",
              "20                D-loop extension by DNA polymerases   \n",
              "21  The Fanconi anemia pathway is required for the...   \n",
              "22  DNA IR-double strand breaks and cellular respo...   \n",
              "23        DNA IR-damage and cellular response via ATR   \n",
              "24                    Glioblastoma signaling pathways   \n",
              "25                              Meiotic recombination   \n",
              "26  Multiple pathways of DNA repair participate in...   \n",
              "27                  DNA repair pathways, full network   \n",
              "28  HDR through Homologous Recombination (HRR) or ...   \n",
              "29  Breast cancer is the leading cause of cancer d...   \n",
              "30                              Breast cancer pathway   \n",
              "31                   Integrated breast cancer pathway   \n",
              "32  The packaging of eukaryotic chromosomal DNA in...   \n",
              "33  DNA double-strand breaks (DSBs) arise in cells...   \n",
              "34                         CHEK1 phosphorylates BRCA2   \n",
              "35               BRCA1 forms a heterodimer with BARD1   \n",
              "36    Some pathogenic BRCA1 mutants do not bind BARD1   \n",
              "37                Defective BARD1 does not bind BRCA1   \n",
              "38                UBXN1 binds BRCA1:BARD1 heterodimer   \n",
              "39          BAP1 disrupts the BRCA1:BARD1 heterodimer   \n",
              "40                 BAP1 binds BRCA1:BARD1 heterodimer   \n",
              "41         BRCA1 gene expression is inhibited by E2F6   \n",
              "42                PIAS1,4 SUMOylates BRCA1 with SUMO1   \n",
              "43              PIAS1,4 SUMOylates BRCA1 with SUMO2,3   \n",
              "44          BRCA1:BARD1 heterodimer autoubiquitinates   \n",
              "45  BRCA1 is a breast and ovarian cancer tumor sup...   \n",
              "46                   BRCA1 binds phosphorylated RBBP8   \n",
              "47      Initial resection of double-strand break ends   \n",
              "48  SMC1/SMC3 role in DNA damage - Cornelia de Lan...   \n",
              "49                           Homologous recombination   \n",
              "50            Resolution of meiotic holliday junction   \n",
              "51  EXO1 or DNA2 in complex with BLM or WRN binds ...   \n",
              "52   Long-range resection of DNA DSBs by EXO1 or DNA2   \n",
              "53  The ataxia telangiectasia-mutated gene (ATM) e...   \n",
              "54  Association of RPA complexes with ssDNA at res...   \n",
              "55  The Wnt pathway plays a critical role in the d...   \n",
              "56                                               None   \n",
              "\n",
              "                                                 LINK TYPE MULTI_N OLAP  \\\n",
              "0   https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "1   https://cgap.nci.nih.gov/Pathways/BioCarta/h_a...    P     252    2   \n",
              "2   https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "3   https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "4   https://reactome.org/PathwayBrowser/#/R-HSA-97...    P   13163    2   \n",
              "5   https://reactome.org/PathwayBrowser/#/R-HSA-97...    P   13163    2   \n",
              "6   https://reactome.org/PathwayBrowser/#/R-HSA-97...    P   13163    2   \n",
              "7   https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "8   https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "9   https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "10  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "11  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "12  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "13  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "14  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "15  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "16  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "17             https://www.genome.jp/pathway/hsa03440    P     345    2   \n",
              "18  https://reactome.org/PathwayBrowser/#/R-HSA-91...    P   13163    2   \n",
              "19  https://reactome.org/PathwayBrowser/#/R-HSA-91...    P   13163    2   \n",
              "20  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    2   \n",
              "21             https://www.genome.jp/pathway/hsa03460    P     345    2   \n",
              "22  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "23  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "24  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "25  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "26  http://www.cs.tau.ac.il/~spike/maps/spike00007...    P      28    2   \n",
              "27  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "28  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "29             https://www.genome.jp/pathway/hsa05224    P     345    2   \n",
              "30  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "31  https://www.wikipathways.org/index.php/Pathway...    P    1333    2   \n",
              "32  http://www.cs.tau.ac.il/~spike/maps/spike00010...    P      28    2   \n",
              "33  http://www.cs.tau.ac.il/~spike/maps/spike00004...    P      28    2   \n",
              "34  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    1   \n",
              "35  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    1   \n",
              "36  https://reactome.org/PathwayBrowser/#/R-HSA-96...    P   13163    1   \n",
              "37  https://reactome.org/PathwayBrowser/#/R-HSA-96...    P   13163    1   \n",
              "38  https://reactome.org/PathwayBrowser/#/R-HSA-97...    P   13163    1   \n",
              "39  https://reactome.org/PathwayBrowser/#/R-HSA-97...    P   13163    1   \n",
              "40  https://reactome.org/PathwayBrowser/#/R-HSA-97...    P   13163    1   \n",
              "41  https://reactome.org/PathwayBrowser/#/R-HSA-90...    P   13163    1   \n",
              "42  https://reactome.org/PathwayBrowser/#/R-HSA-29...    P   13163    1   \n",
              "43  https://reactome.org/PathwayBrowser/#/R-HSA-29...    P   13163    1   \n",
              "44  https://reactome.org/PathwayBrowser/#/R-HSA-97...    P   13163    1   \n",
              "45  https://cgap.nci.nih.gov/Pathways/BioCarta/h_b...    P     252    1   \n",
              "46  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    1   \n",
              "47  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    1   \n",
              "48  https://www.wikipathways.org/index.php/Pathway...    P    1333    1   \n",
              "49  https://www.wikipathways.org/index.php/Pathway...    P    1333    1   \n",
              "50  https://reactome.org/PathwayBrowser/#/R-HSA-91...    P   13163    1   \n",
              "51  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    1   \n",
              "52  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    1   \n",
              "53  https://cgap.nci.nih.gov/Pathways/BioCarta/h_a...    P     252    1   \n",
              "54  https://reactome.org/PathwayBrowser/#/R-HSA-56...    P   13163    1   \n",
              "55  http://www.cs.tau.ac.il/~spike/maps/spike00027...    P      28    1   \n",
              "56             https://www.genome.jp/pathway/hsa05200    P     345    1   \n",
              "\n",
              "        COCO_V2                           SIMILARITY_SCORE        pvalue  \\\n",
              "0    188.558608                                        .52  1.020749e-08   \n",
              "1   5956.815769  .2761904761904761904761904761904761904762  2.143572e-07   \n",
              "2   3370.353336  .2727272727272727272727272727272727272727  2.357930e-07   \n",
              "3   3436.072517  .2666666666666666666666666666666666666667  2.817267e-07   \n",
              "4   3436.072517  .2666666666666666666666666666666666666667  2.817267e-07   \n",
              "5   3436.072517  .2666666666666666666666666666666666666667  2.817267e-07   \n",
              "6   3436.072517  .2666666666666666666666666666666666666667  2.817267e-07   \n",
              "7   3687.288396                                       .264  3.062246e-07   \n",
              "8   3687.288396                                       .264  3.062246e-07   \n",
              "9   3687.288396                                       .264  3.062246e-07   \n",
              "10  3687.288396                                       .264  3.062246e-07   \n",
              "11  3402.224081  .2615384615384615384615384615384615384615  3.317434e-07   \n",
              "12  3402.224081  .2615384615384615384615384615384615384615  3.317434e-07   \n",
              "13  1940.666302  .2571428571428571428571428571428571428571  3.858430e-07   \n",
              "14  3336.772994  .2571428571428571428571428571428571428571  3.858430e-07   \n",
              "15  2709.154452                                        .25  5.062914e-07   \n",
              "16   6125.62014  .2470588235294117647058823529411764705882  5.726401e-07   \n",
              "17  2627.667251  .2390243902439024390243902439024390243902  8.370140e-07   \n",
              "18  115.3844923  .2363636363636363636363636363636363636364  9.656283e-07   \n",
              "19  150.0163697  .2355555555555555555555555555555555555556  1.010541e-06   \n",
              "20   5990.79972  .2333333333333333333333333333333333333333  1.151405e-06   \n",
              "21  2639.032014  .2296296296296296296296296296296296296296  1.460692e-06   \n",
              "22  2707.216113  .2290909090909090909090909090909090909091  1.515812e-06   \n",
              "23  4220.275213                                        .22  3.225566e-06   \n",
              "24  4976.097877   .219277108433734939759036144578313253012  3.473608e-06   \n",
              "25  344.1773602  .2181818181818181818181818181818181818182  3.907426e-06   \n",
              "26   5740.10268  .2156862745098039215686274509803921568627  5.257877e-06   \n",
              "27  6283.961347  .2133333333333333333333333333333333333333  7.288146e-06   \n",
              "28  2430.640086  .2121212121212121212121212121212121212121  8.825394e-06   \n",
              "29   3701.25203   .210884353741496598639455782312925170068  1.095366e-05   \n",
              "30  3126.072387  .2103225806451612903225806451612903225806  1.218264e-05   \n",
              "31  1553.397473  .2101910828025477707006369426751592356688  1.250009e-05   \n",
              "32  1569.958527  .2101910828025477707006369426751592356688  1.250009e-05   \n",
              "33  2707.633139  .2097560975609756097560975609756097560976  1.364333e-05   \n",
              "34  126.2105263  .3666666666666666666666666666666666666667  9.036281e-05   \n",
              "35  126.2105263  .3666666666666666666666666666666666666667  9.036281e-05   \n",
              "36  126.2105263  .3666666666666666666666666666666666666667  9.036281e-05   \n",
              "37  126.2105263  .3666666666666666666666666666666666666667  9.036281e-05   \n",
              "38  384.3965412                                         .3  1.355411e-04   \n",
              "39  134.1614907                                         .3  1.355411e-04   \n",
              "40  134.1614907                                         .3  1.355411e-04   \n",
              "41   756.782076                                        .26  1.807174e-04   \n",
              "42  1216.805852  .2333333333333333333333333333333333333333  2.258917e-04   \n",
              "43  1744.991383  .2142857142857142857142857142857142857143  2.710639e-04   \n",
              "44  1469.189466                                         .2  3.162341e-04   \n",
              "45  2952.437544  .1888888888888888888888888888888888888889  3.614022e-04   \n",
              "46  1317.698072  .1888888888888888888888888888888888888889  3.614022e-04   \n",
              "47  1317.698072  .1888888888888888888888888888888888888889  3.614022e-04   \n",
              "48  1793.045634  .1666666666666666666666666666666666666667  4.968944e-04   \n",
              "49  1401.916365  .1571428571428571428571428571428571428571  5.872123e-04   \n",
              "50  4341.774771  .1571428571428571428571428571428571428571  5.872123e-04   \n",
              "51  2154.832995  .1470588235294117647058823529411764705882  7.226738e-04   \n",
              "52  2154.832995  .1470588235294117647058823529411764705882  7.226738e-04   \n",
              "53  2452.469682  .1421052631578947368421052631578947368421  8.129713e-04   \n",
              "54  3183.379618                                        .14  8.581170e-04   \n",
              "55    1198.7612  .1017817371937639198218262806236080178174  2.003731e-02   \n",
              "56  2740.388822  .1015037593984962406015037593984962406015  2.370460e-02   \n",
              "\n",
              "    Rank      pFDR  \n",
              "0      1  0.000001  \n",
              "1      2  0.000024  \n",
              "2      3  0.000026  \n",
              "3      4  0.000031  \n",
              "4      4  0.000030  \n",
              "5      4  0.000030  \n",
              "6      4  0.000030  \n",
              "7      8  0.000032  \n",
              "8      8  0.000032  \n",
              "9      8  0.000032  \n",
              "10     8  0.000031  \n",
              "11    12  0.000034  \n",
              "12    12  0.000033  \n",
              "13    14  0.000038  \n",
              "14    14  0.000038  \n",
              "15    16  0.000049  \n",
              "16    17  0.000055  \n",
              "17    18  0.000080  \n",
              "18    19  0.000091  \n",
              "19    20  0.000094  \n",
              "20    21  0.000106  \n",
              "21    22  0.000133  \n",
              "22    23  0.000136  \n",
              "23    24  0.000287  \n",
              "24    25  0.000306  \n",
              "25    26  0.000340  \n",
              "26    27  0.000452  \n",
              "27    28  0.000619  \n",
              "28    29  0.000741  \n",
              "29    30  0.000909  \n",
              "30    31  0.000999  \n",
              "31    32  0.001013  \n",
              "32    32  0.001000  \n",
              "33    34  0.001078  \n",
              "34    35  0.007048  \n",
              "35    35  0.006958  \n",
              "36    35  0.006868  \n",
              "37    35  0.006777  \n",
              "38    39  0.010030  \n",
              "39    39  0.009895  \n",
              "40    39  0.009759  \n",
              "41    42  0.012831  \n",
              "42    43  0.015812  \n",
              "43    44  0.018703  \n",
              "44    45  0.021504  \n",
              "45    46  0.024214  \n",
              "46    46  0.023853  \n",
              "47    46  0.023491  \n",
              "48    49  0.031801  \n",
              "49    50  0.036994  \n",
              "50    50  0.036407  \n",
              "51    52  0.044083  \n",
              "52    52  0.043360  \n",
              "53    54  0.047965  \n",
              "54    55  0.049771  \n",
              "55   111  0.040075  \n",
              "56   112  0.023705  "
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "data": {
            "text/plain": [
              "57"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "######\n",
        "# run_pager is a function connected to PAGER api to perform hypergeometric test and retrieve enriched PAGs associated to a list of genes\n",
        "# The input parameters are:\n",
        "# 1.genes: a list of gene symbols.\n",
        "# 2.source: a list of sources refered to 'http://discovery.informatics.uab.edu/PAGER/index.php/pages/help'\n",
        "# 3.Type: a list of PAG types consisting of 'P', 'A' or 'G'\n",
        "# 4.minSize: the allowed minimum size of PAG genes\n",
        "# 5.maxSize: the allowed maximum size of PAG genes\n",
        "# 6.similarity: the similarity score cutoff [0,1]\n",
        "# 7.overlap: the allowed minimum overlap genes\n",
        "# 8.nCoCo: the minimum nCoCo score\n",
        "# 9.pValue: p-value cutoff\n",
        "# 10.FDR: false discovery rate\n",
        "######\n",
        "PAG_result = PAGER.run_pager(\n",
        "    ['BRCA1','BRCA2'],\n",
        "    source = ['KEGG_2021_HUMAN', 'WikiPathway_2021', 'BioCarta', 'Reactome_2021', 'Spike'],\n",
        "    Type='All',\n",
        "    minSize=1,\n",
        "    maxSize=2000,\n",
        "    similarity = 0.05,\n",
        "    overlap = 1,\n",
        "    nCoCo = 0,\n",
        "    pValue = 0.05,\n",
        "    FDR = 0.05\n",
        ")\n",
        "PAG_result\n",
        "len(PAG_result)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 493
        },
        "id": "6DMOfpWpZOtn",
        "outputId": "84fb1ecc-db50-4583-82ee-c2df0677e1c8"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>SYM_A</th>\n",
              "      <th>SYM_B</th>\n",
              "      <th>H_SCORE</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>BARD1</td>\n",
              "      <td>BRCA1</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>BARD1</td>\n",
              "      <td>FANCA</td>\n",
              "      <td>.921</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>BARD1</td>\n",
              "      <td>FANCC</td>\n",
              "      <td>.926</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>BARD1</td>\n",
              "      <td>FANCD2</td>\n",
              "      <td>.9732</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>BARD1</td>\n",
              "      <td>FANCE</td>\n",
              "      <td>.931</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>360</th>\n",
              "      <td>RAD51</td>\n",
              "      <td>RAD9A</td>\n",
              "      <td>.888</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>361</th>\n",
              "      <td>RAD51</td>\n",
              "      <td>TP53</td>\n",
              "      <td>.9754</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>362</th>\n",
              "      <td>RAD9A</td>\n",
              "      <td>TP53</td>\n",
              "      <td>.871</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>363</th>\n",
              "      <td>RAD9A</td>\n",
              "      <td>TREX1</td>\n",
              "      <td>.6804</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>364</th>\n",
              "      <td>TP53</td>\n",
              "      <td>TREX1</td>\n",
              "      <td>.6764</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>365 rows × 3 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "     SYM_A   SYM_B H_SCORE\n",
              "0    BARD1   BRCA1       1\n",
              "1    BARD1   FANCA    .921\n",
              "2    BARD1   FANCC    .926\n",
              "3    BARD1  FANCD2   .9732\n",
              "4    BARD1   FANCE    .931\n",
              "..     ...     ...     ...\n",
              "360  RAD51   RAD9A    .888\n",
              "361  RAD51    TP53   .9754\n",
              "362  RAD9A    TP53    .871\n",
              "363  RAD9A   TREX1   .6804\n",
              "364   TP53   TREX1   .6764\n",
              "\n",
              "[365 rows x 3 columns]"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "data": {
            "text/plain": [
              "<__main__.WeightedGraph at 0x1279a6f20>"
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "best sub: ('BARD1', 'BRCA1', 'ABL1', 'ATM', 'JUN'), max coco: 1.374965299512294\n",
            "best sub: ('BARD1', 'BRCA1', 'ABL1', 'JUN', 'RAD51'), max coco: 2.047478008546958\n",
            "best sub: ('BRCA1', 'FANCA', 'CHEK1', 'BRCA2', 'HUS1'), max coco: 2.8819895403124574\n"
          ]
        },
        {
          "ename": "KeyboardInterrupt",
          "evalue": "",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[6], line 14\u001b[0m\n\u001b[1;32m     10\u001b[0m   pag_graph\u001b[38;5;241m.\u001b[39madd_edge(node1, node2, weight)\n\u001b[1;32m     12\u001b[0m pag_graph\n\u001b[0;32m---> 14\u001b[0m BEST_SUB, MAX_COCO \u001b[38;5;241m=\u001b[39m \u001b[43mfind_max_coco_subgraph\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpag_graph\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     16\u001b[0m BEST_SUB\n\u001b[1;32m     17\u001b[0m MAX_COCO\n",
            "Cell \u001b[0;32mIn[2], line 137\u001b[0m, in \u001b[0;36mfind_max_coco_subgraph\u001b[0;34m(graph)\u001b[0m\n\u001b[1;32m    134\u001b[0m local_triangles \u001b[38;5;241m=\u001b[39m count_triangles_in_subgraph(graph\u001b[38;5;241m.\u001b[39mgraph, subgraph_nodes)\n\u001b[1;32m    136\u001b[0m coco \u001b[38;5;241m=\u001b[39m CoCo(total_nodes, total_interactions, total_triangles, local_nodes, local_interactions, local_triangles)\n\u001b[0;32m--> 137\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[43mcoco\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcoco\u001b[49m \u001b[38;5;241m>\u001b[39m max_score:\n\u001b[1;32m    138\u001b[0m     max_score \u001b[38;5;241m=\u001b[39m coco\u001b[38;5;241m.\u001b[39mcoco\n\u001b[1;32m    139\u001b[0m     best_subgraph \u001b[38;5;241m=\u001b[39m subgraph_nodes\n",
            "Cell \u001b[0;32mIn[2], line 34\u001b[0m, in \u001b[0;36mCoCo.coco\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[38;5;129m@property\u001b[39m\n\u001b[1;32m     33\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mcoco\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m---> 34\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mCoI\u001b[49m \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mCoT\n",
            "Cell \u001b[0;32mIn[2], line 26\u001b[0m, in \u001b[0;36mCoCo.CoI\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[38;5;129m@property\u001b[39m\n\u001b[1;32m     25\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mCoI\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m---> 26\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_compute_score\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkint\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnint\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mKint\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mNint\u001b[49m\u001b[43m)\u001b[49m\n",
            "Cell \u001b[0;32mIn[2], line 21\u001b[0m, in \u001b[0;36mCoCo._compute_score\u001b[0;34m(self, k, n, K, N)\u001b[0m\n\u001b[1;32m     19\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m     20\u001b[0m sign \u001b[38;5;241m=\u001b[39m (k \u001b[38;5;241m/\u001b[39m n) \u001b[38;5;241m-\u001b[39m (K \u001b[38;5;241m/\u001b[39m N)\n\u001b[0;32m---> 21\u001b[0m p \u001b[38;5;241m=\u001b[39m \u001b[43mhypergeom\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpmf\u001b[49m\u001b[43m(\u001b[49m\u001b[43mk\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mN\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mK\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m0\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m p \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mabs\u001b[39m(math\u001b[38;5;241m.\u001b[39mlog10(p)) \u001b[38;5;241m*\u001b[39m (\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m sign \u001b[38;5;241m<\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;241m1\u001b[39m)\n",
            "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/scipy/stats/_distn_infrastructure.py:3392\u001b[0m, in \u001b[0;36mrv_discrete.pmf\u001b[0;34m(self, k, *args, **kwds)\u001b[0m\n\u001b[1;32m   3390\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m np\u001b[38;5;241m.\u001b[39many(cond):\n\u001b[1;32m   3391\u001b[0m     goodargs \u001b[38;5;241m=\u001b[39m argsreduce(cond, \u001b[38;5;241m*\u001b[39m((k,)\u001b[38;5;241m+\u001b[39margs))\n\u001b[0;32m-> 3392\u001b[0m     \u001b[43mplace\u001b[49m\u001b[43m(\u001b[49m\u001b[43moutput\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcond\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclip\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_pmf\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mgoodargs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3393\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m output\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   3394\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m output[()]\n",
            "File \u001b[0;32m<__array_function__ internals>:180\u001b[0m, in \u001b[0;36mplace\u001b[0;34m(*args, **kwargs)\u001b[0m\n",
            "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/numpy/lib/function_base.py:1953\u001b[0m, in \u001b[0;36mplace\u001b[0;34m(arr, mask, vals)\u001b[0m\n\u001b[1;32m   1949\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(arr, np\u001b[38;5;241m.\u001b[39mndarray):\n\u001b[1;32m   1950\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124margument 1 must be numpy.ndarray, \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1951\u001b[0m                     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnot \u001b[39m\u001b[38;5;132;01m{name}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(name\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mtype\u001b[39m(arr)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m))\n\u001b[0;32m-> 1953\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_insert\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mvals\u001b[49m\u001b[43m)\u001b[49m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ],
      "source": [
        "# pagGeneInt is a function connected to PAGER api to retrieve gene interaction network\n",
        "# example:\n",
        "\n",
        "# pagGeneInts = PAGER.pagGeneInt(list(PAG_result.GS_ID)[0])\n",
        "# %debug\n",
        "all_pagGeneInts = pd.concat([PAGER.pagGeneInt(gs_id) for gs_id in list(PAG_result.GS_ID)], ignore_index=True)\n",
        "all_pagGeneInts\n",
        "pag_graph = WeightedGraph()\n",
        "for node1, node2, weight in all_pagGeneInts.values.tolist():\n",
        "  pag_graph.add_edge(node1, node2, weight)\n",
        "\n",
        "pag_graph\n",
        "\n",
        "BEST_SUB, MAX_COCO = find_max_coco_subgraph(pag_graph)\n",
        "\n",
        "BEST_SUB\n",
        "MAX_COCO\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "0    WAG000628\n",
              "1    WAG000644\n",
              "2    WAG000628\n",
              "3    WAG000212\n",
              "Name: GS_A_ID, dtype: object"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "data": {
            "text/plain": [
              "<__main__.WeightedGraph at 0x105d7dd20>"
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        },
        {
          "ename": "TypeError",
          "evalue": "find_max_coco_subgraph() takes 1 positional argument but 2 were given",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[1;32m/Users/qi/Workspace/super_pag/SUPER_PAG_SOLVER.ipynb Cell 7\u001b[0m line \u001b[0;36m1\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/qi/Workspace/super_pag/SUPER_PAG_SOLVER.ipynb#W6sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m   pag_graph\u001b[39m.\u001b[39madd_edge(node1, node2, weight)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/qi/Workspace/super_pag/SUPER_PAG_SOLVER.ipynb#W6sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m pag_graph\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/qi/Workspace/super_pag/SUPER_PAG_SOLVER.ipynb#W6sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m BEST_SUB, MAX_COCO \u001b[39m=\u001b[39m find_max_coco_subgraph(pag_graph, \u001b[39m10\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/qi/Workspace/super_pag/SUPER_PAG_SOLVER.ipynb#W6sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m BEST_SUB\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/qi/Workspace/super_pag/SUPER_PAG_SOLVER.ipynb#W6sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m MAX_COCO\n",
            "\u001b[0;31mTypeError\u001b[0m: find_max_coco_subgraph() takes 1 positional argument but 2 were given"
          ]
        }
      ],
      "source": [
        "# pathInt is a function connected to PAGER api to retrieve the m-type relationships of PAGs using a list of PAG IDs\n",
        "# example:\n",
        "PAGint = PAGER.pathInt(list(PAG_result.GS_ID))\n",
        "PAGint.GS_A_ID\n",
        "pag_graph = WeightedGraph()\n",
        "for node1, node2, weight in zip(PAGint.GS_A_ID, PAGint.GS_B_ID,PAGint.SIMILARITY):\n",
        "  pag_graph.add_edge(node1, node2, weight)\n",
        "\n",
        "pag_graph\n",
        "BEST_SUB, MAX_COCO = find_max_coco_subgraph(pag_graph)\n",
        "BEST_SUB\n",
        "MAX_COCO"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.9"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
